reposerver/repository/repository.go

```diff
@@ -153,84 +182,237 @@ func (s *Service) runRepoOperation(
+			cacheErr := s.cache.GetManifests(revision, q.ApplicationSource, q.Namespace, q.AppLabelKey, q.AppLabelValue, innerRes)
+			if cacheErr != nil && cacheErr != reposervercache.ErrCacheMiss {
+				log.Warnf("manifest cache set error %s: %v", q.ApplicationSource.String(), cacheErr)
+				return nil, cacheErr
+			}
+
+			// If this is the first error we have seen, store the time (we only use the first failure, as this
+			// value is used for PauseGenerationOnFailureForMinutes)
+			if innerRes.FirstFailureTimestamp == 0 {
+				innerRes.FirstFailureTimestamp = s.now().Unix()
+			}
+
+			// Update the cache to include failure information
+			innerRes.NumberOfConsecutiveFailures++
+			innerRes.MostRecentError = err.Error()
+			cacheErr = s.cache.SetManifests(revision, q.ApplicationSource, q.Namespace, q.AppLabelKey, q.AppLabelValue, innerRes)
+			if cacheErr != nil {
+				log.Warnf("manifest cache set error %s: %v", q.ApplicationSource.String(), cacheErr)
+				return nil, cacheErr
+			}
+
 		}
-		res.Revision = revision
-		res.VerifyResult = verifyResult
-		err = s.cache.SetManifests(revision, q.ApplicationSource, q.Namespace, q.AppLabelKey, q.AppLabelValue, &res)
-		if err != nil {
-			log.Warnf("manifest cache set error %s/%s: %v", q.ApplicationSource.String(), revision, err)
+		return nil, err
+	}
+	// Otherwise, no error occurred, so ensure the manifest generation error data in the cache entry is reset before we cache the value
+	manifestGenCacheEntry := cache.CachedManifestResponse{
+		ManifestResponse:                manifestGenResult,
+		NumberOfCachedResponsesReturned: 0,
+		NumberOfConsecutiveFailures:     0,
+		FirstFailureTimestamp:           0,
+		MostRecentError:                 "",
+	}
+	manifestGenResult.Revision = revision
+	manifestGenResult.VerifyResult = verifyResult
+	err = s.cache.SetManifests(revision, q.ApplicationSource, q.Namespace, q.AppLabelKey, q.AppLabelValue, &manifestGenCacheEntry)
+	if err != nil {
+		log.Warnf("manifest cache set error %s/%s: %v", q.ApplicationSource.String(), revision, err)
+	}
+	return manifestGenCacheEntry.ManifestResponse, nil
+}
+
+// getManifestCacheEntry returns false if the 'generate manifests' operation should be run by runRepoOperation, eg:
+// - If the cache result is empty for the requested key
+// - If the cache is not empty, but the cached value is a manifest generation error AND we have not yet met the failure threshold (eg res.NumberOfConsecutiveFailures > 0 && res.NumberOfConsecutiveFailures <  s.initConstants.PauseGenerationAfterFailedGenerationAttempts)
+// - If the cache is not empty, but the cache value is an error AND that generation error has expired
+// and returns true otherwise.
+// If true is returned, either the second or third parameter (but not both) will contain a value from the cache (a ManifestResponse, or error, respectively)
+func (s *Service) getManifestCacheEntry(revision string, inputParam interface{}, firstInvocation bool) (bool, interface{}, error) {
```

Comment:
Looks like `inputParam` is required only to pass `*apiclient.ManifestRequest` to `getManifestCacheEntry` and `runManifestGen` method. We can use closure to archive the same. Also `manifestRunOperationParams` can be removed now, since it has only `*apiclient.ManifestRequest`. 

The https://github.com/alexmt/argo-cd/commit/44bf95ddc15d04974e4eb7a48efee1ff48e17e99 demonstrates 
changes which I'm proposing.

Reply:
Thanks @alexmt, LGTM, accepted as-is and updated the surrounding comments to match.